src/main/java/com/snowflake/kafka/connector/internal/SnowflakeTelemetryPipeStatus.java

```diff
@@ -0,0 +1,181 @@
+  static final String START_TIME    = "start_time";
+  static final String END_TIME      = "end_time";
+
+  SnowflakeTelemetryPipeStatus(final String tableName, final String stageName, final String pipeName)
+  {
+    this.tableName = tableName;
+    this.stageName = stageName;
+    this.pipeName = pipeName;
+
+    this.processedOffset = new AtomicLong(0);
+    this.flushedOffset = new AtomicLong(0);
+    this.committedOffset = new AtomicLong(0);
+    this.purgedOffset = new AtomicLong(0);
+    this.totalNumberOfRecord = new AtomicLong(0);
+    this.totalSizeOfData = new AtomicLong(0);
+    this.fileCountOnStage = new AtomicLong(0);
+    this.fileCountOnIngestion = new AtomicLong(0);
+    this.fileCountPurged = new AtomicLong(0);
+    this.fileCountTableStage = new AtomicLong(0);
+    this.cleanerRestartCount = new AtomicLong(0);
+
+    this.averageKafkaLag = new AtomicLong(0);
+    this.averageKafkaLagRecordCount = new AtomicLong(0);
+    this.averageIngestionLag = new AtomicLong(0);
+    this.averageIngestionLagFileCount = new AtomicLong(0);
+    this.averageCommitLag = new AtomicLong(0);
+    this.averageCommitLagFileCount = new AtomicLong(0);
+    this.startTime = new AtomicLong(System.currentTimeMillis());
+
+    this.lagLock = new ReentrantLock();
+  }
+
+  void updateKafkaLag(final long lag)
+  {
+    updateLag(lag, averageKafkaLagRecordCount, averageKafkaLag);
+  }
+
+  void updateIngestionLag(final long lag)
+  {
+    updateLag(lag, averageIngestionLagFileCount, averageIngestionLag);
+  }
+
+  void updateCommitLag(final long lag)
+  {
+    updateLag(lag, averageCommitLagFileCount, averageCommitLag);
+  }
+
+  private void updateLag(final long lag, AtomicLong averageFileCount, AtomicLong averageLag)
+  {
+    lagLock.lock();
+    try {
+      long lagFileCount = averageFileCount.getAndIncrement();
+      // Weighted average to calculate the cumulative average lag
+      averageLag.updateAndGet(value -> (value * lagFileCount + lag) / (lagFileCount + 1));
+    }
+    finally {
+      lagLock.unlock();
+    }
+  }
+
+  boolean empty()
+  {
+    return this.processedOffset.get() == 0 &&
+            this.flushedOffset.get() == 0 &&
+            this.committedOffset.get() == 0 &&
+            this.purgedOffset.get() == 0 &&
+            this.totalNumberOfRecord.get() == 0 &&
+            this.totalSizeOfData.get() == 0 &&
+            this.fileCountOnStage.get() == 0 &&
+            this.fileCountOnIngestion.get() == 0 &&
+            this.fileCountPurged.get() == 0 &&
+            this.fileCountTableStage.get() == 0 &&
+            this.cleanerRestartCount.get() == 0 &&
+            this.averageKafkaLag.get() == 0 &&
+            this.averageKafkaLagRecordCount.get() == 0 &&
+            this.averageIngestionLag.get() == 0 &&
+            this.averageIngestionLagFileCount.get() == 0 &&
+            this.averageCommitLag.get() == 0 &&
+            this.averageCommitLagFileCount.get() == 0;
+  }
+
+  void dumpTo(ObjectNode msg)
```

Comment:
Seems like `dumpTo` function is a common function of all implementations of `SnowflakeTelemetryBasicInfo`
Can you 
1. change `SnowflakeTelemetryBasicInfo` to `abstract class`
2. move this function to `SnowflakeTelemetryBasicInfo`
3. override `dumpTo` function here